### ISSUE TITLE
Is it possible for Anbox to run without modifying kernel module?

### ISSUE THREAD

**Issue by ghostplant** [anbox/anbox#2098]
I see a project https://github.com/budtmo/docker-android which doesn't require kernel level modules.
What's the problem if Anbox run without building extra kernel module? Because Linux Image is always changing so it makes Anbox's kernel module not able to well compatible with uncertain image version changes.

**Comment by Fuseteam** [comment]
@ghostplant that's because the kernel modules are now part of the linux kernel by default. only certain distros don't enable it ;)

see binder: https://github.com/torvalds/linux/blob/master/drivers/android/Kconfig

**Comment by ghostplant** [comment]
OK, what about Ubuntu 22.04's official kernel, which is based on 5.19.x? Is it integrated by default?

**Comment by Fuseteam** [comment]
22.04 official kernel is based on 5.15.x, it is integrated by default. i have used anbox fine on it.

anything above 5.18 drops ashmem, the other needed kernel module, in favor of memfd. anbox cannot work on those, until the android image is updated.

**Comment by ghostplant** [comment]
Awsome, thanks!

**Comment by ghostplant** [comment]
I checked my Ubuntu 22.04 kernel version which is 5.15.0-48-generic, but I didn't find device file `/dev/ashmem` and `/dev/binder` exist. Is it not a environment that Anbox can directly work?

**Comment by Fuseteam** [comment]
those nodes don't exists like that anymore, try checking with these commands instead
```
grep binder /proc/filesystems
grep ashmem /proc/misc
```
as i said, i have run anbox on ubuntu 22.04 just fine without messing with the kernel

**Comment by ghostplant** [comment]
So strange, the word doesn't exist as well:
```sh
root@ubuntu-pc:~# grep ashmem /proc/misc
root@ubuntu-pc:~# grep binder /proc/filesystems
root@ubuntu-pc:~# uname -a
Linux ubuntu-pc 5.15.0-48-generic #54-Ubuntu SMP Fri Aug 26 13:26:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
root@ubuntu-pc:~# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.1 LTS"
```

**Comment by Fuseteam** [comment]
Hmm
Try 
```
sudo modprobe binder_linux
sudo modprobe ashmem_linux
```

**Comment by ghostplant** [comment]
OK, it works by installing anbox from snapd. There is a snapshot, some area are covered with black, is it as expected, or a display bug?
![anbox](https://user-images.githubusercontent.com/12099308/195284660-0c048d6c-67dc-475d-93fb-a458cee3421f.png)

**Comment by Fuseteam** [comment]
a known display bug, no known solution

**Comment by ghostplant** [comment]
@Fuseteam BTW, does `houdini_7y/8y/9y` work for Anbox? I always get error code 255 by running arm-based executable.

**Comment by Fuseteam** [comment]
Houdini isn't shipped, you'll need to add it

**Comment by ghostplant** [comment]
Yes, I mean I configured that, and I also have tried both 7y and 9y, but basic arm-x32 executable never works. Without configuring houdini, running executables returns "error ELF executable format" which is reasonable, after houdini is configured, the error message goes, and it just directly exit with 255 error code.
```sh
$ ./busybox.arm ls
255 $ true
0 $
```

I'm not sure whether there is any difference between Anbox and Androidx86 OS in configuring houdini. I ever configured the houdini in 32-bit Android x86 OS using houdini_7x, which worked well for busybox.arm.

**Comment by ghostplant** [comment]
> Houdini isn't shipped, you'll need to add it

I just found the problem is because I don't have root access in adb shell, even though after `su` to become root user. So the command `mount -t binfmt_misc none /proc/sys/fs/binfmt_misc` failed inside adb for permission deny.

How can I manage to mount it successfully? I can mount it successfully at host side, so is there any way I can bind host-side `/proc/sys/fs/binfmt_misc` into android's lxc container?

**Comment by Fuseteam** [comment]
> Yes, I mean I configured that, and I also have tried both 7y and 9y, but basic arm-x32 executable never works. Without configuring houdini, running executables returns "error ELF executable format" which is reasonable, after houdini is configured, the error message goes, and it just directly exit with 255 error code.
> 
> ```shell
> $ ./busybox.arm ls
> 255 $ true
> 0 $
> ```
> 
> I'm not sure whether there is any difference between Anbox and Androidx86 OS in configuring houdini. I ever configured the houdini in 32-bit Android x86 OS using houdini_7x, which worked well for busybox.arm.

oh interesting, this might be because anbox doesn't support 32bit so perhaps that extends to arm emulation

> I just found the problem is because I don't have root access in adb shell, even though after `su` to become root user. So the command `mount -t binfmt_misc none /proc/sys/fs/binfmt_misc` failed inside adb for permission deny.

ah nevermind, ignore me

> How can I manage to mount it successfully? I can mount it successfully at host side, so is there any way I can bind host-side `/proc/sys/fs/binfmt_misc` into android's lxc container?

not quite sure, anbox is fairly isolated to the point it has absolutely no hardware access, aside from sound and keyboard/mouse input. so perhaps you need an "pipe" to enable that or something, sadly that's outside my knowledge set

**Comment by ghostplant** [comment]
OK, I just installed an x86-64 based Chrome.apk, but it just crashed sadly. This is something well supported by Android-x86 OS, not using houdini.

**Comment by Fuseteam** [comment]
that could be related to the outdate android image, right now it's based on android 7

**Comment by ghostplant** [comment]
I do use android 7: https://build.anbox.io/android-images/2018/07/19/android_amd64.img
The error reported "X GL error", then the appmgr window closed.

**Comment by Fuseteam** [comment]
No i mean anbox __is__ using android 7, chrome.apk might no longer support it
